src/twister2/generate_tests_plugin.py

```diff
@@ -0,0 +1,146 @@
+"""
+Plugin generates variants of tests bases on specification from YAML file.
+Test variants are generated for `platform` and `scenario`.
+"""
+
+from __future__ import annotations
+
+import itertools
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+
+import pytest
+
+from twister2.exceptions import TwisterConfigurationException
+from twister2.helper import safe_load_yaml
+from twister2.platform_specification import PlatformSpecification
+from twister2.yaml_file import extract_tests, should_be_skip
+from twister2.yaml_test_specification import YamlTestSpecification
+
+TEST_SPEC_FILE_NAME = 'testspec.yaml'
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Variant:
+    """Keeps information about single test variant"""
+    platform: PlatformSpecification
+    scenario: str
+
+    def __str__(self):
+        return f'{self.platform.identifier}:{self.scenario}'
+
+
+def get_scenarios_from_fixture(metafunc: pytest.Metafunc) -> list[str]:
+    """Return scenarios selected by fixture `build_specification`."""
+    if mark := metafunc.definition.get_closest_marker('build_specification'):
+        if build_name := mark.kwargs.get('names'):
+            return build_name.split(',')
+        else:
+            logger.warning('Found `build_specification` decorator, but without `names` of configuration')
+    return []
+
+
+def get_scenarios_from_yaml(spec_file: Path) -> list[str]:
+    """Return all available scenarios from yaml specification."""
+    data = safe_load_yaml(spec_file)
+    try:
+        return data['tests'].keys()
+    except KeyError:
+        return []
+
+
+def pytest_generate_tests(metafunc: pytest.Metafunc):
+    # generate parametrized tests for each selected platform for ordinary pytest tests
+    # if `specification` fixture is used
+    if 'specification' not in metafunc.fixturenames:
+        return
+
+    twister_config = metafunc.config.twister_config
+
+    platforms_list: list[PlatformSpecification] = [
+        platform for platform in twister_config.platforms
+        if platform.identifier in twister_config.default_platforms
+    ]
+    spec_file_path: Path = Path(metafunc.definition.fspath.dirname) / TEST_SPEC_FILE_NAME
+    scenarios = get_scenarios_from_fixture(metafunc)
+    if not scenarios:
+        scenarios = get_scenarios_from_yaml(spec_file_path)
+    variants = itertools.product(platforms_list, scenarios)
+    params: list[pytest.param] = []
+    for variant in variants:
+        v = Variant(*variant)
+        params.append(
+            pytest.param(v, marks=pytest.mark.platform(v.platform.identifier), id=str(v))
+        )
+
+    # using indirect=True to inject value from `specification` fixture instead of param
+    metafunc.parametrize(
+        'specification', params, scope='function', indirect=True
+    )
+
+
+@pytest.fixture(scope='function')
+def specification(request: pytest.FixtureRequest) -> None:
```

Comment:
I am not sure if the name of this fixture is suitable. Maybe different name like `variant` would be better.

Reply:
In the end it will return a build output, right? Hence, I think having a "build" in the name might be helpful. Maybe `build_variant`, or `build_item`?

Reply:
It returns `None`, but `pytest.mark.parametrize` use this fixture to inject class `twister.generate_test_plugin.Variant` to test function. It is used to generate parametrized tests for different platforms and scenarios. 
The body of `specification` is empty due to all necessary work is done under the hood, but we need to keep this fixture to have information which tests contain yaml file with a build specification (there may be some tests without yaml spec).